### PR TITLE
Improve/bands definition

### PR DESCRIPTION
**Pull request by Joao-L-S-Almeida**
It adds support to a easier way to define the dataset and output bands when there are many variables.
Now the bands can be defined by intervals:
```
output_bands:
 -  [0,10]

```
Or subintervals:
```
output_bands:
 - [0,5]
 - [5,10]
 
```
See `tests/manufactured-finetune_prithvi_swin_B_band_interval.yaml` for more details. 

Furthermore, the bands also can be defined via strings, as:

```
dataset_bands:
 -  "band_0"
 -  "band_1"
 -  "band_2"
 -  "band_3"
 
 output_bands:
 - "band_1"
 - "band_2"
```

### CODE REVIEW

File: terratorch/datamodules/generic_pixel_wise_data_module.py [hunk 1]
```diff
@@ -91,9 +91,9 @@ def __init__(
         test_split: Path | None = None,
         ignore_split_file_extensions: bool = True,
         allow_substring_split_file: bool = True,
-        dataset_bands: list[HLSBands | int] | None = None,
-        predict_dataset_bands: list[HLSBands | int] | None = None,
-        output_bands: list[HLSBands | int] | None = None,
+        dataset_bands: list[HLSBands | int | list[int] | str] | None = None,
+        predict_dataset_bands: list[HLSBands | int | list[int] | str ] | None = None,
+        output_bands: list[HLSBands | int | list[int] | str] | None = None,
```
Comment (CarlosGomes98) [inline terratorch/datamodules/generic_pixel_wise_data_module.py:96]: instead of `list[int]` would it work if we made it a `tuple[int, int]`? In that way we would have a check that exactly 2 elements are there

File: terratorch/datamodules/generic_pixel_wise_data_module.py [hunk 2]
```diff
@@ -330,9 +330,9 @@ def __init__(
         test_split: Path | None = None,
         ignore_split_file_extensions: bool = True,
         allow_substring_split_file: bool = True,
-        dataset_bands: list[HLSBands | int] | None = None,
-        predict_dataset_bands: list[HLSBands | int] | None = None,
-        output_bands: list[HLSBands | int] | None = None,
+        dataset_bands: list[HLSBands | int | list[int] | str ] | None = None,
+        predict_dataset_bands: list[HLSBands | int | list[int] | str ] | None = None,
+        output_bands: list[HLSBands | int | list[int] | str ] | None = None,
```
Comment (CarlosGomes98) [inline terratorch/datamodules/generic_pixel_wise_data_module.py:335]: same as above

File: terratorch/datasets/generic_pixel_wise_dataset.py [hunk 3]
```diff
@@ -43,8 +43,8 @@ def __init__(
         ignore_split_file_extensions: bool = True,
         allow_substring_split_file: bool = True,
         rgb_indices: list[int] | None = None,
-        dataset_bands: list[HLSBands | int] | None = None,
-        output_bands: list[HLSBands | int] | None = None,
+        dataset_bands: list[HLSBands | int | list[int] | str ] | None = None,
+        output_bands: list[HLSBands | int | list[int] | str ] | None = None,
```
Comment (CarlosGomes98) [inline terratorch/datasets/generic_pixel_wise_dataset.py:47]: same as above

File: terratorch/datasets/generic_pixel_wise_dataset.py [hunk 4]
```diff
@@ -120,19 +121,38 @@ def __init__(
             )
         self.rgb_indices = [0, 1, 2] if rgb_indices is None else rgb_indices
 
-        self.dataset_bands = dataset_bands
-        self.output_bands = output_bands
+        bands_by_interval = (self._bands_defined_by_interval(bands_list=dataset_bands) and
+                             self._bands_defined_by_interval(bands_list=output_bands))
```
Comment (CarlosGomes98) [inline terratorch/datasets/generic_pixel_wise_dataset.py:125]: The function _bands_defined_by_interval doing all these checks is not super pythonic. Could we not just have a function that goes through all the elements of `dataset_bands` and `output_bands` and, if they are an interval, expands them into the full list? (what `_generate_bands_intervals` does).

Then we avoid needing these checks because after this we are in a common format with no ranges.

File: terratorch/datasets/generic_pixel_wise_dataset.py [hunk 5]
```diff
@@ -154,18 +174,56 @@ def __getitem__(self, index: int) -> dict[str, Any]:
...
         if nan_replace is not None:
             data = data.fillna(nan_replace)
         return data
 
+    def _generate_bands_intervals(self, bands_intervals:List[List[int]] = None):
+        bands = list()
+        for b_interval in bands_intervals:
+            bands_sublist = np.arange(b_interval[0], b_interval[1] + 1).astype(int).tolist()
+            bands.append(bands_sublist)
+        return sorted(sum(bands, []))
```
Comment (CarlosGomes98) [inline terratorch/datasets/generic_pixel_wise_dataset.py:195]: why sorted? Doesn't this destroy the order that the user requests?

e.g. If my `output_bands` are `[6-10]`, `[2, 4]`, doesn't this return `[2, 3, 4, 6, 7, 8, 9, 10]`? I would expect `[6, 7, 8, 9, 10, 2, 3, 4]`

**Review by CarlosGomes98 — COMMENTED** [review]
Just some initial comments, have not been able to finish the review yet

**Comment by CarlosGomes98** [comment]
@Joao-L-S-Almeida I like the idea of the ranges! However, in this idea, I am wondering if it makes sense to still retain `dataset_bands`.

Maybe what makes most sense is to give users two options:

1. Use `dataset_bands` to name your bands using strings or `HLSBands` (or ints?) and then `output_bands` will match those.
2. Use only `output_bands` without naming the bands with `dataset_bands`, in which case we can then accept ints or ranges as a tuple of two ints.

What caught my attention was that, when using ranges, having to specify that same range in `dataset_bands` is sort of pointless

What do you think?

**Comment by CarlosGomes98** [comment]
Also, lets make sure to add documentation for this new feature! (extending the docstrings is probably enough)

File: terratorch/datasets/generic_pixel_wise_dataset.py [hunk 6]
```diff
@@ -348,8 +406,8 @@ def __init__(
         ignore_split_file_extensions: bool = True,
         allow_substring_split_file: bool = True,
         rgb_indices: list[int] | None = None,
-        dataset_bands: list[HLSBands | int] | None = None,
-        output_bands: list[HLSBands | int] | None = None,
+        dataset_bands: list[HLSBands | int | list[int] | str ] | None = None,
+        output_bands: list[HLSBands | int | list[int] | str ] | None = None,
```
Comment (CarlosGomes98) [inline terratorch/datasets/generic_pixel_wise_dataset.py:410]: same as above

File: terratorch/datasets/generic_pixel_wise_dataset.py [hunk 7]
```diff
@@ -120,19 +121,38 @@ def __init__(
...
+            self.output_bands = self._generate_bands_intervals(output_bands)
+        else:
+            self.dataset_bands = dataset_bands
+            self.output_bands = output_bands
+            
+            bands_type = self._bands_as_int_or_str(dataset_bands, output_bands)
+
+            if bands_type == str:
+                UserWarning("When the bands are defined as str, guarantee your input files"+ 
+                                  "are organized by band and all have its specific name.")
```
Comment (CarlosGomes98) [inline terratorch/datasets/generic_pixel_wise_dataset.py:139]: I don't think this is worth a warning, the user could be doing everything correctly but would receive a warning every time they run their code. We should instead make it clear in the documentation / docstring exactly what each case expects (str vs range and int)

File: terratorch/datasets/generic_pixel_wise_dataset.py (hunk 5 above)
Comment (CarlosGomes98) [reply terratorch/datasets/generic_pixel_wise_dataset.py:195]: Also, no need for `np` and then `tolist()` here. Consider `list(range(b_interval[0], b_interval[1] + 1))`

File: terratorch/datasets/generic_pixel_wise_dataset.py [hunk 8]
```diff
@@ -154,18 +174,56 @@ def __getitem__(self, index: int) -> dict[str, Any]:
...
-
+    
     def _load_file(self, path, nan_replace: int | float | None = None) -> xr.DataArray:
         data = rioxarray.open_rasterio(path, masked=True)
         if nan_replace is not None:
             data = data.fillna(nan_replace)
         return data
 
+    def _generate_bands_intervals(self, bands_intervals:List[List[int]] = None):
+        bands = list()
```
Comment (CarlosGomes98) [inline terratorch/datasets/generic_pixel_wise_dataset.py:191]: Prefer `bands = []`

File: terratorch/datasets/generic_pixel_wise_dataset.py (hunk 5 above)
Comment (Joao-L-S-Almeida) [reply terratorch/datasets/generic_pixel_wise_dataset.py:195]: Really, will remove it.

File: tests/manufactured-finetune_prithvi_swin_B_band_interval.yaml [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+      - class_path: albumentations.Rotate
+        init_args:
+          limit: 30
+          border_mode: 0 # cv2.BORDER_CONSTANT
+          value: 0
+          # mask_value: 1
+          p: 0.5
+      - class_path: ToTensorV2
+    dataset_bands:
+      - [0, 11]
```
[the remark this answers is not in the document]
Comment (Joao-L-S-Almeida) [reply tests/manufactured-finetune_prithvi_swin_B_band_interval.yaml:46]: The only advantage I see to keep dataset_bands is inform someone reading the yaml file how many bands were in the dataset used as input.

File: terratorch/datamodules/generic_pixel_wise_data_module.py (hunk 1 above)
Comment (Joao-L-S-Almeida) [reply terratorch/datamodules/generic_pixel_wise_data_module.py]: 0a84c42e0975fcc0a57740f4dc6b3340d3e956a2

File: terratorch/datamodules/generic_pixel_wise_data_module.py (hunk 2 above)
Comment (Joao-L-S-Almeida) [reply terratorch/datamodules/generic_pixel_wise_data_module.py]: 0a84c42e0975fcc0a57740f4dc6b3340d3e956a2

File: terratorch/datasets/generic_pixel_wise_dataset.py (hunk 3 above)
Comment (Joao-L-S-Almeida) [reply terratorch/datasets/generic_pixel_wise_dataset.py]: 0a84c42e0975fcc0a57740f4dc6b3340d3e956a2

File: terratorch/datasets/generic_pixel_wise_dataset.py [hunk 10]
```diff
@@ -181,8 +239,8 @@ def __init__(
         ignore_split_file_extensions: bool = True,
         allow_substring_split_file: bool = True,
         rgb_indices: list[str] | None = None,
-        dataset_bands: list[HLSBands | int] | None = None,
-        output_bands: list[HLSBands | int] | None = None,
+        dataset_bands: list[HLSBands | int | list[int] | str ] | None = None,
+        output_bands: list[HLSBands | int | list[int] | str ] | None = None,
```
[the remark this answers is not in the document]
Comment (Joao-L-S-Almeida) [reply terratorch/datasets/generic_pixel_wise_dataset.py]: 0a84c42e0975fcc0a57740f4dc6b3340d3e956a2

File: terratorch/datasets/generic_pixel_wise_dataset.py (hunk 6 above)
Comment (Joao-L-S-Almeida) [reply terratorch/datasets/generic_pixel_wise_dataset.py]: 0a84c42e0975fcc0a57740f4dc6b3340d3e956a2

File: terratorch/datasets/generic_pixel_wise_dataset.py [hunk 11]
```diff
@@ -120,19 +121,38 @@ def __init__(
...
+
+        # If the bands are defined by sub-intervals or not.
+        if bands_by_interval:
+            self.dataset_bands = self._generate_bands_intervals(dataset_bands)
+            self.output_bands = self._generate_bands_intervals(output_bands)
+        else:
+            self.dataset_bands = dataset_bands
+            self.output_bands = output_bands
+            
+            bands_type = self._bands_as_int_or_str(dataset_bands, output_bands)
```
[the remark this answers is not in the document]
Comment (Joao-L-S-Almeida) [reply terratorch/datasets/generic_pixel_wise_dataset.py]: 303bfe89a22a20cbe65de90fbf4d701aff7df25f

File: terratorch/datasets/generic_pixel_wise_dataset.py (hunk 7 above)
Comment (Joao-L-S-Almeida) [reply terratorch/datasets/generic_pixel_wise_dataset.py]: 303bfe89a22a20cbe65de90fbf4d701aff7df25f

File: terratorch/datasets/generic_pixel_wise_dataset.py [hunk 12]
```diff
@@ -120,19 +121,38 @@ def __init__(
...
+
+            if bands_type == str:
+                UserWarning("When the bands are defined as str, guarantee your input files"+
```
[the remark this answers is not in the document]
Comment (Joao-L-S-Almeida) [reply terratorch/datasets/generic_pixel_wise_dataset.py]: 303bfe89a22a20cbe65de90fbf4d701aff7df25f

File: terratorch/datasets/generic_pixel_wise_dataset.py [hunk 13]
```diff
@@ -120,19 +121,38 @@ def __init__(
...
+                             self._bands_defined_by_interval(bands_list=output_bands))
+
+        # If the bands are defined by sub-intervals or not.
+        if bands_by_interval:
+            self.dataset_bands = self._generate_bands_intervals(dataset_bands)
+            self.output_bands = self._generate_bands_intervals(output_bands)
+        else:
+            self.dataset_bands = dataset_bands
+            self.output_bands = output_bands
+
```
[the remark this answers is not in the document]
Comment (Joao-L-S-Almeida) [reply terratorch/datasets/generic_pixel_wise_dataset.py]: Reformatted with black.

File: terratorch/datasets/generic_pixel_wise_dataset.py [hunk 14]
```diff
@@ -154,18 +174,56 @@ def __getitem__(self, index: int) -> dict[str, Any]:
...
             "filename": self.image_files[index],
         }
+
         if self.reduce_zero_label:
             output["mask"] -= 1
         if self.transform:
             output = self.transform(**output)
         return output
-
+
```
[the remark this answers is not in the document]
Comment (Joao-L-S-Almeida) [reply terratorch/datasets/generic_pixel_wise_dataset.py:176]: Reformatted with black.

File: terratorch/datasets/generic_pixel_wise_dataset.py [hunk 15]
```diff
@@ -154,18 +174,56 @@ def __getitem__(self, index: int) -> dict[str, Any]:
...
+        band_type = [None, None]
+        if not dataset_bands and not output_bands:
+            return None
+        else:
+            for b, bands_list in enumerate([dataset_bands, output_bands]):
+                if all([type(band)==int for band in bands_list]):
+                    band_type[b] = int
+                elif all([type(band)==str for band in bands_list]):
+                    band_type[b] = str
```
[the remark this answers is not in the document]
Comment (Joao-L-S-Almeida) [reply terratorch/datasets/generic_pixel_wise_dataset.py:200]: If the inputs are `None`, as seen for `tests/test_generic_dataset.py`, it produces an error inside the list-comprehension.

File: terratorch/datasets/generic_pixel_wise_dataset.py (hunk 5 above)
Comment (Joao-L-S-Almeida) [reply terratorch/datasets/generic_pixel_wise_dataset.py]: 10551a0d59295101575540925bc6e49bc276e3fc

File: terratorch/datasets/generic_pixel_wise_dataset.py (hunk 8 above)
Comment (Joao-L-S-Almeida) [reply terratorch/datasets/generic_pixel_wise_dataset.py]: 10551a0d59295101575540925bc6e49bc276e3fc

File: terratorch/datasets/generic_pixel_wise_dataset.py [hunk 16]
```diff
@@ -154,18 +174,56 @@ def __getitem__(self, index: int) -> dict[str, Any]:
...
             data = data.fillna(nan_replace)
         return data
 
+    def _generate_bands_intervals(self, bands_intervals:List[List[int]] = None):
+        bands = list()
+        for b_interval in bands_intervals:
+            bands_sublist = np.arange(b_interval[0], b_interval[1] + 1).astype(int).tolist()
+            bands.append(bands_sublist)
+        return sorted(sum(bands, []))
+
```
[the remark this answers is not in the document]
Comment (Joao-L-S-Almeida) [reply terratorch/datasets/generic_pixel_wise_dataset.py]: 10551a0d59295101575540925bc6e49bc276e3fc

File: tests/test_finetune.py [hunk 17]
```diff
@@ -23,21 +23,29 @@ def test_finetune_multiple_backbones(model_name):
...
+    model_instance = timm.create_model(model_name)
+
+    state_dict = model_instance.state_dict()
+
+    torch.save(state_dict, os.path.join("tests/", model_name + ".pt"))
+
+    # Running the terratorch CLI
+    command_list = ["fit", "-c", f"tests/manufactured-finetune_{model_name}_string.yaml"]
+    _ = build_lightning_cli(command_list)
+
```
[the remark this answers is not in the document]
Comment (Joao-L-S-Almeida) [reply tests/test_finetune.py:51]: Right. See commit: 5708e0a3f433b999c2c4fce520a52d4593cfa942

**Review by CarlosGomes98 — APPROVED** [review]
